### ISSUE TITLE
Material gets recreated on every render loop

### ISSUE THREAD

**Issue by gromgull** [FarazzShaikh/THREE-CustomShaderMaterial#12]
This thing is amazing - but I've noticed the material gets recreated each loop - even if the props are the same :( 

if you add a `console.log(material.current)` to your water example it should happen there too.

I am not sure how, but I believe the problem is here: https://github.com/FarazzShaikh/THREE-CustomShaderMaterial/blob/master/src/index.tsx#L18-L22

If I add the material to my tree as `customerShaderMaterial` and pass the args array the way the constructor expects it works.

**Comment by FarazzShaikh** [comment]
Huh this sucks, I will look into this

**Comment by gromgull** [comment]
I solved it by creating the material in a `useAsset` block - caching it "globally" - then passing it in as an attribute `<mesh material={myMaterial} ... ` and not inline as a jsx element. But it's not as nice.

**Comment by FarazzShaikh** [comment]
Yeah you could do the same with useMemo, I’ll be refactoring this library again with the patterns I learnt working on other projects to make it solid and easy to work with.

**Comment by FarazzShaikh** [comment]
Im not sure where to add `console.log(material.current)` to test this. Do you mean to `useFrame`? 

I added it outside `useFrame` and It's only recreated every time the component is re-rendered. Adding it inside useFrame shows that the UUID of the material remains unchanged, which implies it is not re-initialized on each renader

**Comment by FarazzShaikh** [comment]
OK so I added a log to the constructor of the CustomShaderMaterial class to see how many times its initialized.

The water example currently initializes it twice, however, moving the `<Water />` component outside of `<Suspense />` makes it only initialize once. I have pushed the "fixed" example to the repo and sandbox.

Suspense might be forcing a re-render of everything within it. Does this solve it?
